File: cpp/ubuntu/anders/DataSet.cpp
#include "DataSet.h"

void DataSet::SetValue(int rowNum, string fieldName, string fieldValue)
{
	ROW row;
	DATA_SET::iterator it = ds.find(rowNum);
	if (it == ds.end())
	{
		ds.insert(pair<int, ROW>(rowNum, row));
	}
	else
	{
		row = it->second;
	}
	
	row.insert(pair<string, string>(fieldName, fieldValue));
}

string DataSet::GetValue(int rowNum, string fieldName)
{
	string returnValue;
	ROW row;
	DATA_SET::iterator it = ds.find(rowNum);
	if (it != ds.end())
	{
		ROW::iterator rowIt = it->second.find(fieldName);
		if (rowIt != it->second.end())
			returnValue = rowIt->second;
	}

	return returnValue;
}

int DataSet::GetSize()
{
	if (ds.empty())
		return -1;
	else
		return ds.size();
}

File: delphi/hhre/dataset.pas
unit dataset;

interface

uses
  Generics.Collections, constant, SysUtils;

type
  TDataSet = class
  strict private
    ds: TDictionary<Integer, TDictionary<string, string>>;
  public
    constructor Create;
    procedure SetValue(const rowNum: Integer; const fieldName, fieldValue: string);
    function getValue(const rowNum: Integer; const fieldName: string): string;
    function getSize: Integer;
  end;

implementation

constructor TDataSet.Create;
begin
  Self.ds := TDictionary<Integer, TDictionary<string, string>>.Create;
end;

function TDataSet.getSize: Integer;
begin
  if ds = nil then
    Result := RETURN_ZERO
  else
    Result := ds.Count;
end;

function TDataSet.getValue(const rowNum: Integer; const fieldName: string): string;
var
  row: TDictionary<string, string>;
  returnValue: string;
begin
  Result := EmptyStr;
  if Self.ds.TryGetValue(rowNum, row) then
    if row.TryGetValue(fieldName, returnValue) then
      Result := returnValue;
end;

procedure TDataSet.SetValue(const rowNum: Integer; const fieldName, fieldValue: string);
var
  row: TDictionary<string, string>;
begin
  if not Self.ds.TryGetValue(rowNum, row) then
  begin
    row := TDictionary<string, string>.Create;
    ds.Add(rowNum, row);
  end;

  row.Add(fieldName, fieldValue);
end;

end.
